crd-generator/api/src/main/java/io/fabric8/crd/generator/DeterministicObjectMapper.java

```diff
@@ -0,0 +1,206 @@
+ * limitations under the License.
+ */
+package io.fabric8.crd.generator;
+
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.fasterxml.jackson.databind.ser.std.StdDelegatingSerializer;
+import com.fasterxml.jackson.databind.util.Converter;
+import com.fasterxml.jackson.databind.util.StdConverter;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_EMPTY;
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
+import static com.fasterxml.jackson.annotation.JsonInclude.Value.construct;
+
+/**
+ * A {@link ObjectMapper} that serializes YAML in a deterministic way.
+ *
+ * Converts unsorted collections into sorted lists, using the provided converters.
+ */
+class DeterministicObjectMapper {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DeterministicObjectMapper.class);
+
+  static DeterministicObjectMapperBuilder builder() {
+    return new DeterministicObjectMapperBuilder();
+  }
+
+  static class DeterministicObjectMapperBuilder {
+
+    private final CustomComparators customComparators = new CustomComparators();
+
+    private DeterministicObjectMapperBuilder() {
+    }
+
+    <T, U extends Comparable<? super U>> DeterministicObjectMapperBuilder withConverter(final Class<T> clazz,
+        final Function<? super T, ? extends U> keyExtractor) {
+      return withConverter(clazz, Comparator.comparing(keyExtractor));
+    }
+
+    <T> DeterministicObjectMapperBuilder withConverter(final Class<T> clazz, final Comparator<T> comparator) {
+      customComparators.addConverter(clazz, comparator);
+      return this;
+    }
+
+    ObjectMapper build() {
+      final ObjectMapper mapper = JsonMapper.builder(new YAMLFactory()
+          .enable(YAMLGenerator.Feature.MINIMIZE_QUOTES)
+          .enable(YAMLGenerator.Feature.ALWAYS_QUOTE_NUMBERS_AS_STRINGS)
+          .disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
+          .configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true)
```

Comment:
Have you attempted to only add those configurations: `MapperFeature.SORT_PROPERTIES_ALPHABETICALLY` and `SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS` to the old mapper without defining an entirely new and custom one?
What are the invariants that we are missing?

Reply:
The problem is that a lot of relevant elements of a CRD are not maps, but sets or other unsorted collections. The custom code is only taking care of collections.

`MapperFeature.SORT_PROPERTIES_ALPHABETICALLY` and `SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS` is already configured in my code, but it won't affect structures like enums, the additional printer columns section etc.

Reply:
With all tests of the `crd-generator` the following collections are serialized:
```
com.fasterxml.jackson.databind.node.TextNode
io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceColumnDefinition
io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinitionVersion
io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceColumnDefinition
io.fabric8.kubernetes.api.model.apiextensions.v1.JSONSchemaProps
```
I've added custom converters for all of them, including v1 and v1beta1 variants.